ygen/codegen.go

```diff
@@ -367,6 +393,101 @@ func (cg *YANGCodeGenerator) GenerateGoCode(yangFiles, includePaths []string) (*
 	}, nil
 }
 
+// GenerateProto3 generates Protobuf 3 code for the input set of YANG files.
+// The YANG schemas for which protobufs are to be created is supplied as the
+// yangFiles argument, with included modules being searched for in includePaths.
+// Returns a GeneratedProto3 struct containing the messages that are to be
+// output, along with any associated values (e.g., enumerations).
+func (cg *YANGCodeGenerator) GenerateProto3(yangFiles, includePaths []string) (*GeneratedProto3, *YANGCodeGeneratorError) {
+	// TODO(robjs): Handle enumerated types in proto messages.
+	msgs, _, st, errs := langAgnosticDefinitions(yangFiles, includePaths, cg.Config)
+	if len(errs) > 0 {
+		return nil, &YANGCodeGeneratorError{Errors: errs}
+	}
+
+	cg.state.schematree = st
+
+	protoMsgs, errs := cg.state.buildDirectoryDefinitions(msgs, cg.Config.CompressOCPaths, cg.Config.GenerateFakeRoot, protobuf)
+
+	if len(errs) > 0 {
+		return nil, &YANGCodeGeneratorError{Errors: errs}
+	}
+
+	p := &GeneratedProto3{
+		Packages: map[string]Proto3Package{},
+	}
+	ye := NewYANGCodeGeneratorError()
+
+	// Write out the protobuf messages in a determinstic order.
+	msgMap := map[string]*yangDirectory{}
+	msgNames := []string{}
+	for _, m := range protoMsgs {
+		msgNames = append(msgNames, m.name)
+		msgMap[m.name] = m
+	}
+	sort.Strings(msgNames)
+
+	// pkgImports lists the imports that are required for the package that is being
+	// written out.
+	pkgImports := map[string][]string{}
+
+	for _, n := range msgNames {
+		m := msgMap[n]
+		pkg, msg, reqs, errs := writeProtoMsg(m, protoMsgs, cg.state, cg.Config.CompressOCPaths)
+
+		if _, ok := pkgImports[pkg]; !ok {
+			pkgImports[pkg] = []string{}
+		}
+
+		for _, i := range reqs {
+			var found bool
+			for _, e := range pkgImports[pkg] {
+				if i == e {
+					found = true
+				}
+			}
+			if !found {
+				pkgImports[pkg] = append(pkgImports[pkg], i)
+			}
+		}
+
+		if len(errs) > 0 {
+			ye.Errors = append(ye.Errors, errs...)
+		}
+
+		tp, ok := p.Packages[pkg]
+		if !ok {
+			p.Packages[pkg] = Proto3Package{
+				Messages: []string{},
+			}
+			tp = p.Packages[pkg]
+		}
+		tp.Messages = append(tp.Messages, msg)
+		p.Packages[pkg] = tp
+	}
+
+	// Ensure that the base import path is a valid directory.
```

Comment:
Suggest turning this into a helper function (e.g. reformatDir) and implementing in terms of filepath e.g. 
return filepath.Join(strings.Split(path, "/")...)
You could add the trailing slash or (better) continue to use filepath.Join wherever the dir is used. 
Also, the comment suggests an error check so it may be better to say something like "Reformat the directory path into canonical form".

Reply:
Done - added template helper functions that do this.